lib/block_view.go

```diff
@@ -1628,27 +1628,49 @@ func (bav *UtxoView) DisconnectBlock(
+	// Stake distribution UtxoOps may be either an AddBalance or a StakeDistribution operation type.
+	// We need to revert these before iterating over the transactions in the block.
 	if desoBlock.Header.Height >= uint64(bav.Params.ForkHeights.BalanceModelBlockHeight) {
 		if len(utxoOps) != len(desoBlock.Txns)+1 {
 			return fmt.Errorf(
 				"DisconnectBlock: Expected number of utxo ops to be equal to number of txns in block plus one for"+
 					" delete expired nonces operation for block %d",
 				desoBlock.Header.Height)
 		}
-		// We need to revert the delete expired nonces operation.
-		deleteExpiredNoncesUtxoOps := utxoOps[len(utxoOps)-1]
-		if deleteExpiredNoncesUtxoOps[0].Type != OperationTypeDeleteExpiredNonces {
-			return fmt.Errorf(
-				"DisconnectBlock: Expected last utxo op to be delete expired nonces operation for block %d",
-				desoBlock.Header.Height)
-		}
-		if len(deleteExpiredNoncesUtxoOps) != 1 {
-			return fmt.Errorf(
-				"DisconnectBlock: Expected exactly utxo op for deleting expired nonces operation for block %d",
-				desoBlock.Header.Height)
-		}
-		for _, nonceEntry := range deleteExpiredNoncesUtxoOps[0].PrevNonceEntries {
-			bav.SetTransactorNonceEntry(nonceEntry)
+		var isLastBlockInEpoch bool
+		isLastBlockInEpoch, err = bav.IsLastBlockInCurrentEpoch(desoBlock.Header.Height)
+		if err != nil {
+			return errors.Wrapf(err, "DisconnectBlock: Problem checking if block is last in epoch")
+		}
+		blockLevelUtxoOps := utxoOps[len(utxoOps)-1]
+		for ii := len(blockLevelUtxoOps) - 1; ii >= 0; ii-- {
+			utxoOp := blockLevelUtxoOps[ii]
+			switch utxoOp.Type {
+			case OperationTypeDeleteExpiredNonces:
+				// We need to revert the delete expired nonces operation.
+				for _, nonceEntry := range utxoOp.PrevNonceEntries {
+					bav.SetTransactorNonceEntry(nonceEntry)
+				}
+			case OperationTypeAddBalance:
+				// We don't allow add balance utxo operations unless it's the end of an epoch.
+				if !isLastBlockInEpoch {
+					return fmt.Errorf("DisconnectBlock: Found add balance operation in block %d that is not the end of an epoch", desoBlock.Header.Height)
+				}
+				// We need to revert the add balance operation.
+				if err = bav._unAddBalance(utxoOp.BalanceAmountNanos, utxoOp.BalancePublicKey); err != nil {
+					return errors.Wrapf(err, "DisconnectBlock: Problem unAdding balance %v: ", utxoOp.BalanceAmountNanos)
+				}
+			case OperationTypeStakeDistribution:
+				// We don't allow stake distribution utxo operations unless it's the end of an epoch.
+				if !isLastBlockInEpoch {
+					return fmt.Errorf("DisconnectBlock: Found add balance operation in block %d that is not the end of an epoch", desoBlock.Header.Height)
+				}
+				if len(utxoOp.PrevStakeEntries) != 1 {
+					return fmt.Errorf("DisconnectBlock: Expected exactly one prev stake entry for stake distribution op")
+				}
+				bav._setStakeEntryMappings(utxoOp.PrevStakeEntries[0])
```

Comment:
I think you're not resetting the validator amount nanos or the global staked amount nanos here :)

Reply:
We no longer have a global staked amount nanos tracker actually! adding validator reset in another PR.